src/main/java/com/vmware/vipclient/i18n/messages/api/opt/local/LocalLocaleOpt.java

```diff
@@ -37,72 +39,104 @@ public LocalLocaleOpt(LocaleDTO dto) {
 	}
 
 	@Override
-    public Map<String, String> getSupportedLanguages(String locale) {
+    public void getSupportedLanguages(String locale, LocaleCacheItem cacheItem) {
 		Map<String, String> supportedLanguageNames = new HashMap<String, String>();
 		List<String> supportedLanguages = DataSourceEnum.Bundle.createProductOpt(dto)
 				.getSupportedLocales();
 		if(supportedLanguages != null && !supportedLanguages.isEmpty()) {
 			Map<String, String> languagesNames = getLanguagesNamesFromCLDR(locale);
 			if (languagesNames == null || languagesNames.isEmpty())
-				return supportedLanguageNames;
+				return;
 			for(String language : supportedLanguages){
 				if(!ConstantsKeys.SOURCE.equalsIgnoreCase(language))
 					supportedLanguageNames.put(language, (String) languagesNames.get(language));
 			}
 		}
-		return supportedLanguageNames;
+		if (!supportedLanguageNames.isEmpty()) {
+			cacheItem.addCachedData(supportedLanguageNames);
+			cacheItem.setTimestamp(System.currentTimeMillis());
+		}
     }
 
     private Map<String, String> getLanguagesNamesFromCLDR(String locale){
-		Map<String, String> languagesNames = null;
+		LocaleCacheItem languagesNames = null;
 		logger.debug("Look for languages' names from cache for locale [{}]!", locale);
 		FormattingCacheService formattingCacheService = new FormattingCacheService();
 		languagesNames = formattingCacheService.getLanguagesNames(locale);// key
 		if (languagesNames != null) {
+			if (languagesNames.isExpired()) { // cacheItem has expired
```

Comment:
@jessiejuachon Your above 2 points are correct.

In online mode, when fetching supported languages from data source, I needn't do anything except sending request to '/i18n/api/v2/locale/supportedLanguageList' API in Singleton service to get the result, since all the logic is handled by Singleton service;

but in offline mode, I need implement the logic as the same as Singleton service. 
Firstly need get the supported languageTags;
Secondly get the languages' name of the requested locale;
Then get the supported languageTags' name.

Since Singleton service use the same pattern jar with Java client, and it caches the languages' names, so I cache the languages' names in client, too.

Though the code is there, it may not be executed. Because in online&mix mode, supported languages are get from Singleton service; in offline mode, the result of condition 'languagesNames.isExpired()' is false.

This is left just for future extension when local bundle can be updated at runtime.

Reply:
First, local bundle will never be updated at runtime. Reason is this is against product integrity. Any updates to local bundles must be done using an official product upgrade/pathc/hotfix. 

Second, the implementation that you copied from Singleton service (caching the language names) is not optimal. You do not need to copy it. We have to remove it from here, especially because adding the languages' names to the cache is going to unnecessarily add to the memory footprint of the client application, not just in Singleton service. We should eventually remove this non-optimal behavior from the Singleton service too.

Moreover, you are incorrect when you say that the code will not be used. The expire logic in formatCacheItem.isExpired will always be applied in mixed mode, and so the formatCacheItem in cache will expire even if it previously came from the offline jar. To avoid this, we can just remove the caching for language names entirely: 

- FormattingCacheItem.addLanguagesNames
- FormattingCacheItem.getLanguagesNames
- FormattingCacheItem.getLanguagesNamesCacheKey
- ConstantsKeys.LANGUAGES_PREFIX
-

Reply:
First, local bundle will never be updated at runtime. The reason is this is against product integrity. Any updates to local bundles must be done using an official product upgrade/patch/hotfix. 

Second, the implementation that you copied from Singleton service (caching the language names) is not optimal. You do not need to copy it. We have to remove it from here, especially because adding the languages' names to the cache is going to unnecessarily add to the memory footprint of the client application, not just in Singleton service. Removing the caching logic for language names entirely will avoid this problem. We should eventually remove this same non-optimal behavior from the Singleton service too.

Moreover, you are incorrect when you say that the code will not be used. The expire logic in formatCacheItem.isExpired will always be applied in mixed mode, and so the formatCacheItem in cache will expire even if it previously came from the offline jar. Anyway, this is not the primary concern, but removing the caching logic for language names entirely will avoid this problem.

Hence, please remove:
- FormattingCacheItem.addLanguagesNames
- FormattingCacheItem.getLanguagesNames
- FormattingCacheItem.getLanguagesNamesCacheKey
- ConstantsKeys.LANGUAGES_PREFIX

Reply:
First, local bundle will never be updated at runtime. The reason is this is against product integrity. Any updates to local bundles must be done using an official product upgrade/patch/hotfix. 

Second, the implementation that you copied from Singleton service (caching the language names) is not optimal. You do not need to copy it. We have to remove it from here, especially because adding the languages' names to the cache is going to unnecessarily add to the memory footprint of the client application, not just in Singleton service. Removing the caching logic for language names entirely will avoid this problem. We should eventually remove this same non-optimal behavior from the Singleton service too.

Moreover, you are incorrect when you say that the code will not be used. The expire logic in formatCacheItem.isExpired will always be applied in mixed mode, and so the formatCacheItem in cache will expire even if it previously came from the offline jar. Anyway, this is not the primary concern, but removing the caching logic for language names entirely will avoid this problem.

Hence, please remove:
- FormattingCacheItem.addLanguagesNames
- FormattingCacheItem.getLanguagesNames
- FormattingCacheItem.getLanguagesNamesCacheKey
- ConstantsKeys.LANGUAGES_PREFIX
- Other related code